File: tools/improve/tests/original/maudec/py/simple-nested-rewrite.py
import enum

class Bar(enum.Enum):
	A = 0  # 
	B = 1  # 
	C = 2  # 


class Foo(enum.Enum):
	F = 0  # Bar, Foo


class Ter(enum.Enum):
	G = 0  # Bar


def rlapp_Bar(term: Bar) -> Bar | None:
	match term:
		case (Bar.A, ):
			return (Bar.B, )

		case (Bar.B, ):
			return (Bar.C, )

		case _:
			return None



def arlapp_Bar(term: Bar) -> Bar | None:
	return rlapp_Bar(term)


def arlapp_Foo(term: Foo) -> Foo | None:
	match term:
		case (Foo.F, t0, t1):
			if (subt := arlapp_Bar(t0)) is not None:
				return (Foo.F, subt, t1)


			if (subt := arlapp_Foo(t1)) is not None:
				return (Foo.F, t0, subt)




	return None

File: tools/improve/tests/original/maudec/rust/simple-nested-rewrite.rs


#[derive(Clone, PartialEq, PartialOrd, Debug)]
enum Bar {
	A,
	B,
	C,
}

#[derive(Clone, PartialEq, PartialOrd, Debug)]
enum Foo {
	F(Bar, Box<Foo>),
}

#[derive(Clone, PartialEq, PartialOrd, Debug)]
enum Ter {
	G(Bar),
}

fn rlapp_Bar(term: &Bar) -> Option<Bar> {
	match term {
		Bar::A => Some(Bar::B),
		Bar::B => Some(Bar::C),
		_ => None,
	}
}

fn arlapp_Bar(term: &Bar) -> Option<Bar> {
	rlapp_Bar(term)
}

fn arlapp_Foo(term: &Foo) -> Option<Foo> {
	match term {
		Foo::F(t0, t1) => {
			if let Some(subt) = arlapp_Bar(t0) {
				return Some(Foo::F(subt, (*t1).clone()));
			}

			if let Some(subt) = arlapp_Foo(&**t1) {
				return Some(Foo::F((*t0).clone(), subt.into()));
			}

		},
	}	None
}

